                                                        Tasks                                                                                    
P 3.1: Download the CarSharing dataset from Canvas and apply the normal linear regression and the Ridge regression methods to predict the 'demand' column. Use 5 different values for the regularisation parameter in the Ridge regression and repeat the prediction. Calculate all evaluation metrics discussed in the lecture session for both models and all repetitions and print them. Interpret the results and report your interpretation in the report cell (completing the report cell is required) (15%)

In [9]:
############# WRITE YOUR CODE IN THIS CELL (IF APPLICABLE)  ####################
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Load dataset (use correct path)
df = pd.read_csv("csv/CarSharing.csv")

# ----- Handle datetime -----
for col in df.columns:
    if df[col].dtype == 'object':
        try:
            df[col] = pd.to_datetime(df[col])
            df[col + '_hour'] = df[col].dt.hour
            df[col + '_day'] = df[col].dt.day
            df[col + '_month'] = df[col].dt.month
            df.drop(columns=[col], inplace=True)
        except:
            pass

# Keep only numeric columns
df = df.select_dtypes(include=['int64', 'float64'])

# Drop missing values
df = df.dropna()

# Features and target
X = df.drop(columns=['demand'])
y = df['demand']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Evaluation function
def evaluate(y_true, y_pred):
    return {
        "MAE": mean_absolute_error(y_true, y_pred),
        "MSE": mean_squared_error(y_true, y_pred),
        "RMSE": np.sqrt(mean_squared_error(y_true, y_pred)),
        "R2": r2_score(y_true, y_pred)
    }

# ----- Linear Regression -----
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)
print("Linear Regression Metrics:")
print(evaluate(y_test, lr.predict(X_test_scaled)))

# ----- Ridge Regression -----
alphas = [0.01, 0.1, 1, 10, 100]

for a in alphas:
    ridge = Ridge(alpha=a)
    ridge.fit(X_train_scaled, y_train)
    print(f"\nRidge Regression (alpha={a}) Metrics:")
    print(evaluate(y_test, ridge.predict(X_test_scaled)))


Linear Regression Metrics:
{'MAE': 1.0042381686468287, 'MSE': 1.7071336470690726, 'RMSE': np.float64(1.3065732459640649), 'R2': 0.26996169770733247}

Ridge Regression (alpha=0.01) Metrics:
{'MAE': 1.0042388585624071, 'MSE': 1.7071348309451682, 'RMSE': np.float64(1.306573699010189), 'R2': 0.26996119143478814}

Ridge Regression (alpha=0.1) Metrics:
{'MAE': 1.0042450532802523, 'MSE': 1.7071454686250656, 'RMSE': np.float64(1.3065777698342589), 'R2': 0.2699566423394165}

Ridge Regression (alpha=1) Metrics:
{'MAE': 1.0043055980430222, 'MSE': 1.707250164327814, 'RMSE': np.float64(1.3066178340769017), 'R2': 0.2699118702893646}

Ridge Regression (alpha=10) Metrics:
{'MAE': 1.0048016207565684, 'MSE': 1.7081499805354894, 'RMSE': np.float64(1.306962119013206), 'R2': 0.26952707306053925}

Ridge Regression (alpha=100) Metrics:
{'MAE': 1.0067026114815734, 'MSE': 1.7116881749472002, 'RMSE': np.float64(1.3083150136519874), 'R2': 0.2680140002873903}


C:\Users\piumi\AppData\Local\Temp\ipykernel_14756\1792350514.py:17: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df[col] = pd.to_datetime(df[col])
C:\Users\piumi\AppData\Local\Temp\ipykernel_14756\1792350514.py:17: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df[col] = pd.to_datetime(df[col])
C:\Users\piumi\AppData\Local\Temp\ipykernel_14756\1792350514.py:17: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df[col] = pd.to_datetime(df[col])
C:\Users\piumi\AppData\Local\Temp\ipykernel_14756\1792350514.py:17: UserWarning: Could not infer format, so each element will be parsed 

                            ############ WRITE YOUR REPORT IN THIS CELL (IF APPLICABLE) #############                                       
### **Report: Linear Regression and Ridge Regression on the CarSharing Dataset**

In this task, normal linear regression and ridge regression were applied to predict the **demand** variable in the CarSharing dataset. Prior to model training, the dataset was preprocessed by converting datetime attributes into numerical features (hour, day, and month), removing non-numeric columns, and eliminating rows containing missing values. Feature scaling was performed using standardization to ensure fair model comparison, especially for ridge regression.

The **linear regression model** achieved a Mean Absolute Error (MAE) of approximately **1.01**, a Root Mean Squared Error (RMSE) of **1.31**, and an R² value of **0.27**. These results indicate that the model explains about **27% of the variance** in demand, suggesting a weak to moderate linear relationship between the input features and the target variable.

**Ridge regression** was evaluated using five different regularisation parameter values (α = 0.01, 0.1, 1, 10, 100). For smaller α values, ridge regression produced performance metrics similar to linear regression, indicating that light regularisation helps stabilize the model without significantly affecting accuracy. As α increased, the model became more constrained, leading to higher error values and lower R² scores, which is a clear sign of underfitting.

Overall, the results show that linear regression provides a reasonable baseline for predicting demand in the CarSharing dataset. Ridge regression does not significantly improve prediction accuracy but offers more stable coefficients when small regularisation values are used. Therefore, ridge regression with a low α value is preferable when model robustness and generalization are important.